README
XE(1)                       General Commands Manual                      XE(1)

NAME
     xe – execute a command for every argument

SYNOPSIS
     xe [-0FLRnqv] [-I replace-arg] [-N maxargs] [-j maxjobs] command ...
     xe [flags ...] -p pattern command ... [+ pattern command ...]...
     xe [flags ...] -f argfile command ...
     xe [flags ...] -s shellscript
     xe [flags ...] -a command ... -- args ...
     xe [flags ...] -A argsep command ... argsep args ...

DESCRIPTION
     The xe utility constructs command lines from specified arguments,
     combining some of the best features of xargs(1) and apply(1).

     xe means “execute for every ...”.

     xe supports different methods to specify arguments to commands:

     command ...
             By default, arguments - separated by newlines - are read from the
             standard input.  The resulting command is constructed from the
             command line parameters, replacing replace-arg with the read
             argument, and is executed with execvp(3).

             In this mode, no shell is involved and replace-arg must appear as
             a word on its own, i.e.  ‘foo {} bar’ will work, but ‘foo{} bar’
             will not, where {} is the default value for replace-arg.

             If no argument is specified, the default is ‘printf %s\n’.

     -f argfile
             Read arguments from argfile, instead of the standard input.

             This does not close the standard input for execution, it is
             passed to the forked process.

     -s shellscript
             In this mode, the single parameter shellscript is executed using
             sh -c.  In the script, the specified arguments can be accessed
             using $1, $2, ...

             For example:
                   echo 'a\nb' | xe -N2 -s 'echo $2 $1'

     -a command ... -- args ...
             In this mode, everything after -- is passed as args to command.

     -A argsep command ... argsep args ...
             Same as -a, but the custom argument separator argsep is used to
             distinguish between command and its args.

     The options are as follows:

     -0      Input arguments are separated by NUL bytes (instead of newlines,
             which is the default).

     -F      Fatal: stop and exit when a command execution fails.

     -L      Run the resulting commands with line-buffered output; lines from
             two jobs will not interleave.  When used twice, or with -vv, also
             prefix each line with the number of the job (see ENVIRONMENT) in
             such a manner that the output can be piped to ‘sort -snk1’ to
             group it.

     -R      Return with status 122 when no arguments have been specified
             (instead of 0, the default).  xe never executes a command when no
             arguments are specified.

     -n      Dry run: don't run the resulting commands, just print them.

     -q      Quiet mode: redirect standard output and standard error of
             commands to /dev/null.

     -v      Verbose: print commands to standard error before running them.
             When used twice, also print job id and exit status for each
             command.

     -p      Enable make(1)-style percent rules.  The first argument of
             command ... is regarded as a pattern, see PERCENT RULES below.
             Patterns without a slash (or ‘**’) are matched against the
             basenames only.

             Multiple runs of patterns and commands are separated by ‘+’.
             Only the first matching percent rule is executed; in case no
             pattern matches, no command is run.

     -I replace-arg
             Replace first occurrence of replace-arg (default: {}) in the
             resulting command with the argument(s).  Pass an empty
             replace-arg to disable the replace function.  Contrary to
             xargs(1) this will expand into multiple arguments when needed.

     -N maxargs
             Pass up to maxargs arguments to each command (default: 1).
             Using -N0 will pass as many arguments as possible.

     -j maxjobs
             Run up to maxjobs processes concurrently.  Using -j0 will run as
             many processes as there are CPU cores running.  If maxjobs ends
             with an ‘x’, it is regarded as a multiplier of the number of
             running CPU cores (rounded down, but using at least one core).

PERCENT RULES
     The percent rules of xe are similar to the globs of sh(1) or fnmatch(3):
     ‘?’ matches a single character that is not ‘/’.  ‘/’ matches one or
     multiple ‘/’ in the string.  ‘*’ matches zero or more characters, but
     never ‘/’.  ‘**’ matches zero or more characters, including ‘/’.  Note
     that all of these also match leading dots in file names.

     ‘{a,b,c}’ matches either a, b or c.  ‘[abc]’ matches one of the
     characters abc (but never ‘/’).  ‘[!abc]’ matches all characters but abc.
     Alternatively, ‘[^abc]’ can be used too.  ‘[a-c]’ matches any character
     in the range between a and c inclusive.  In character ranges, characters
     can be escaped using a backslash.

     In the pattern, a single occurrence of ‘%’ matches one or more
     characters, and replaces the first occurrence of ‘%’ with the matched
     string in the remaining arguments, which are then used as the command to
     be executed.

ENVIRONMENT
     The environment variable ITER is passed to the child process and
     incremented on each command execution.

EXIT STATUS
     xe follows the convention of GNU and OpenBSD xargs:
     0       on success
     123     if any invocation of the command exited with status 1 to 254.
     124     if the command exited with status 255
     125     if the command was killed by a signal
     126     if the command cannot be run
     127     if the command was not found
     1       if some other error occurred

     Additionally, 122 is returned when -R was passed and the command was
     never executed.

EXAMPLES
     Compress all .c files in the current directory, using all CPU cores:
           xe -a -j0 gzip -- *.c
     Remove all empty files, using lr(1):
           lr -U -t 'size == 0' | xe -N0 rm
     Convert .mp3 to .ogg, using all CPU cores:
           xe -a -j0 -s 'ffmpeg -i "${1}" "${1%.mp3}.ogg"' -- *.mp3
     Same, using percent rules:
           xe -a -j0 -p %.mp3 ffmpeg -i %.mp3 %.ogg -- *.mp3
     Similar, but hiding output of ffmpeg, instead showing spawned jobs:
           xe -ap -j0 -vvq '%.{m4a,ogg,opus}' ffmpeg -y -i {} out/%.mp3 -- *

SEE ALSO
     apply(1), parallel(1), xapply(1), xargs(1)

AUTHORS
     Leah Neukirchen <leah@vuxu.org>

LICENSE
     xe is in the public domain.

     To the extent possible under law, the creator of this work has waived all
     copyright and related or neighboring rights to this work.

     http://creativecommons.org/publicdomain/zero/1.0/

Void Linux                      August 1, 2023                      Void Linux


NEWS.md
## HEAD

## 1.0 (2023-08-05)

* Slightly incompatible change: properly forward errors on exec,
  return 123 on any exit status 1-254 (like GNU xargs).
* Small bugfixes.

## 0.11 (2017-11-05)

* Add zsh completion.
* Fix `-F` to exit as soon as possible when a command failed.
* Fix when using `-s` and `-p` together.
* Better error handling for some rare situations.
* Made test suite more robust.

## 0.10 (2017-10-30)

* New flag `-p` to enable "percent rules" ala make(1), see the man
  page for details.
* New flag `-q` to run commands, hiding their output.
* Fix issues launching commands on FreeBSD and DragonFlyBSD.

## 0.9 (2017-07-17)

* New flag `-L` to enable line buffering on output.
* New flag `-LL` to prefix output lines with `$ITER`.
* `-j` now can scale with the available CPU cores
  (e.g. `-j0.5x` to use half the cores).
* `-vv` will trace `$ITER` and `exit status`.
* Fix issues related to having more children than we knew of.
* Documentation fixes and improvement by Larry Hynes.

## 0.8 (2017-06-02)

* Only read new arguments when we immediately need them

## 0.7 (2017-04-13)

* Error when the `-A` separator does not appear
* Bugfix: fix quoting of empty strings
* Documentation fixes


xa.c
/*
 * xa - simple xargs and apply replacement
 *
 * To the extent possible under law,
 * Christian Neukirchen <chneukirchen@gmail.com>
 * has waived all copyright and related or neighboring rights to this work.
 * http://creativecommons.org/publicdomain/zero/1.0/
 */

#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>

static char delim = '\n';
static char default_replace[] = "{}";
static char *replace = default_replace;
static char default_argsep[] = "--";
static char *argsep = default_argsep;
static char **args;
static char *sflag;

static int maxatonce = 1;
static int maxjobs = 1;
static int runjobs = 0;
static int aflag, nflag, vflag;

static char *
xstrdup(const char *s)
{
	char *d = strdup(s);
	if (!d)
		exit(1);
	return d;
}

static char *getarg_line = 0;
static size_t getarg_len = 0;

static char *
getarg()
{
	if (aflag) {
		if (args && *args)
			return xstrdup(*args++);
		else
			return 0;
	}

	int read = getdelim(&getarg_line, &getarg_len, delim, stdin);
	if (read == -1) {
		if (feof(stdin))
			return 0;
		else
			exit(1);
	}
	if (getarg_line[read-1] == delim)  // strip delimiter
		getarg_line[read-1] = 0;

	return xstrdup(getarg_line);
}

static int
mywait()
{
	int status;
	pid_t pid;

	pid = wait(&status);
	if (pid < 0) {
		if (errno == ECHILD)
			return 0;
		// no other error possible?
	}	
	
	if (WIFEXITED(status)) {
		if (WEXITSTATUS(status) >= 1 && WEXITSTATUS(status) <= 125) {
			exit(123);
		} else if (WEXITSTATUS(status) == 255) {
			fprintf(stderr, "xa: %d exited with status 255\n", pid);
			exit(124);
		} else if (WEXITSTATUS(status) > 125) {
			exit(WEXITSTATUS(status));
		}
	} else if (WIFSIGNALED(status)) {
		fprintf(stderr, "xa: %d terminated by signal %d\n",
		    pid, WTERMSIG(status));
		exit(125);
	}
	
	runjobs--;
	return 1;
}

static void
shquote(const char *s)
{
	if (!strpbrk(s, "\001\002\003\004\005\006\007\010"
	                "\011\012\013\014\015\016\017\020"
	                "\021\022\023\024\025\026\027\030"
	                "\031\032\033\034\035\036\037\040"
	                "`^#*[]=|\\?${}()'\"<>&;\127")) {
		printf("%s", s);
		return;
	}

	putchar('\'');
	for (; *s; s++)
		if (*s == '\'')
			printf("'\\''");
		else
			putchar(*s);
	putchar('\'');
}

static int
trace(char *cmd[])
{
	int i;
	
	for (i = 0; cmd[i]; i++) {
		if (i > 0)
			printf(" ");
		shquote(cmd[i]);
	}
	printf("\n");

	return 0;
}

static int
run(char *cmd[])
{
	pid_t pid;
	int i;

	if (runjobs >= maxjobs)
		mywait();
	runjobs++;

	if (vflag || nflag)
		trace(cmd);
	if (nflag) {
		runjobs--;
		return 0;
	}

	pid = fork();
	if (pid == 0) {  // in child
		// redirect stdin to /dev/null
		int fd = open("/dev/null", O_RDONLY);
		if (fd >= 0) {
			if (dup2(fd, 0) != 0)
				exit(1);
			close(fd);
			execvp(cmd[0], cmd);
		}
		fprintf(stderr, "xa: %s: %s\n", cmd[0], strerror(errno));
		exit(errno == ENOENT ? 127 : 126);
	}

	if (pid < 0)
		exit(126);

	for (i = 0; cmd[i]; i++)
		free(cmd[i]);
	
	return 0;
}

int
main(int argc, char *argv[])
{
	char c;
	int i, cmdend;
	char *arg, **cmd;

	while ((c = getopt(argc, argv, "+0A:I:N:anj:s:v")) != -1)
		switch(c) {
		case '0': delim = '\0'; break;
		case 'A': argsep = optarg; aflag++; break;
                case 'I': replace = optarg; break;
		case 'N': maxatonce = atoi(optarg); break;
		case 'a': aflag++; break;
		case 'n': nflag++; break;
		case 'j': maxjobs = atoi(optarg); break;
		case 's': sflag = optarg; break;
		case 'v': vflag++; break;
		default:
			fprintf(stderr, 
			    "Usage: %s [-0nv] [-I arg] [-N maxargs] [-j maxjobs] COMMAND...\n"
			    "     | -s SHELLSCRIPT\n"
			    "     | -a COMMAND... -- ARGS...\n"
			    "     | -A ARGSEP COMMAND... ARGSEP ARGS...\n",
			    argv[0]);
			exit(1);
		}

	cmdend = argc;
	if (aflag)  // find argsep in argv
		for (i = optind; i < argc; i++)
			if (strcmp(argv[i], argsep) == 0) {
				args = argv + i+1;
				cmdend = i;
				break;
			}

	cmd = calloc(argc-optind+maxatonce+1+
	    (optind==cmdend ? 2 : 0)+(sflag ? 4 : 0), sizeof (char *));
	if (!cmd)
		exit(1);

	while ((arg = getarg())) {
		int l = 0;

		if (sflag) {
			cmd[l++] = xstrdup("/bin/sh");
			cmd[l++] = xstrdup("-c");
			cmd[l++] = xstrdup(sflag);
			cmd[l++] = xstrdup("-");
		}

		if (optind == cmdend) {
			cmd[l++] = xstrdup("printf");
			cmd[l++] = xstrdup("%s\\n");
		}

		if (maxatonce == 1) {
			// substitute {}
			int substituted = 0;
			for (i = optind; i < cmdend; i++) {
				if (strcmp(argv[i], replace) == 0) {
					cmd[l++] = arg;
					substituted = 1;
				} else {
					cmd[l++] = xstrdup(argv[i]);
				}
			}
			if (!substituted)
				cmd[l++] = arg;
		} else {
			// just append to cmd
			for (i = optind; i < cmdend; i++)
				cmd[l++] = xstrdup(argv[i]);
			cmd[l++] = arg;
			for (i = 0; i < maxatonce - 1; i++) {
				cmd[l] = getarg();
				if (!cmd[l++])
					break;
			}
		}
		cmd[l] = 0;
		run(cmd);
	}

	while (mywait())
		;

	free(cmd);
	free(getarg_line);
	return 0;
}


xe.c
/*
 * xe - simple xargs and apply replacement
 *
 * To the extent possible under law, Leah Neukirchen <leah@vuxu.org>
 * has waived all copyright and related or neighboring rights to this work.
 * http://creativecommons.org/publicdomain/zero/1.0/
 */

#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static char delim = '\n';
static char default_replace[] = "{}";
static char *replace = default_replace;
static char *argsep;
static char *fflag;
static char *sflag;

static int maxatonce = 1;
static int maxjobs = 1;
static int runjobs = 0;
static int failed = 0;
static int Aflag, Fflag, Lflag, Rflag, aflag, nflag, pflag, qflag, vflag;
static long iterations = 0;
static FILE *traceout;
static FILE *input;

static size_t argmax;

static char *buf;
static size_t buflen;
static size_t bufcap;

static char **args;
static size_t argslen;
static size_t argscap;
static size_t argsresv;

struct child {
	pid_t pid;  // 0 if empty slot
	long iter;
};
static struct child *children;

static char **inargs;

static char *line = 0;
static size_t linelen = 0;

static char *
getarg()
{
	if (aflag || Aflag) {
		if (inargs && *inargs)
			return *inargs++;
		else
			return 0;
	}

	int read = getdelim(&line, &linelen, delim, input);
	if (read == -1) {
		if (feof(input))
			return 0;
		else
			exit(1);
	}
	if (line[read-1] == delim)  // strip delimiter
		line[read-1] = 0;

	return line;
}

static int
mywait()
{
	int status;
	pid_t pid;

	pid = wait(&status);
	if (pid < 0)
		return !(errno == ECHILD);

	int i;
	for (i = 0; i < maxjobs; i++)
		if (children[i].pid == pid)
			goto my_child;

	return 1;

my_child:
	if (WIFEXITED(status)) {
		if (WEXITSTATUS(status) == 255) {
			fprintf(stderr, "xe: job %ld [pid %ld] exited with status 255\n", children[i].iter, (long)pid);
			exit(124);
		} else if (WEXITSTATUS(status) > 0) {
			if (Fflag) {
				fprintf(stderr, "xe: job %ld [pid %ld] exited with status %d\n", children[i].iter, (long)pid, WEXITSTATUS(status));
				exit(123);
			}
			failed = 1;
		}
	} else if (WIFSIGNALED(status)) {
		fprintf(stderr, "xe: job %ld [pid %ld] terminated by signal %d\n",
		    children[i].iter, (long)pid, WTERMSIG(status));
		exit(125);
	}

	if (vflag > 1) {
		fprintf(traceout, "%04ld< status %d\n",
		    children[i].iter, WEXITSTATUS(status));
		fflush(traceout);
	}

	children[i].pid = 0;
	runjobs--;

	return 1;
}

static void
sh_quote(const char *s)
{
	if (*s &&
	    !strpbrk(s, "\001\002\003\004\005\006\007\010"
	                "\011\012\013\014\015\016\017\020"
	                "\021\022\023\024\025\026\027\030"
	                "\031\032\033\034\035\036\037\040"
	                "`^#*[]=|\\?${}()'\"<>&;\177")) {
		fprintf(traceout, "%s", s);
		return;
	}

	fprintf(traceout, "\'");
	for (; *s; s++)
		if (*s == '\'')
			fprintf(traceout, "'\\''");
		else
			fprintf(traceout, "%c", *s);
	fprintf(traceout, "\'");
}

static int
trace()
{
	size_t i;

	for (i = 0; i < argslen; i++) {
		if (i > 0)
			fprintf(traceout, " ");
		sh_quote(args[i]);
	}
	fprintf(traceout, "\n");
	fflush(traceout);

	return 0;
}

static void
scanargs()
{
	char *s = buf;
	size_t i;

	for (i = 0; i < argslen; i++) {
		args[i] = s;
		s += strlen(s) + 1;
	}
	args[i] = 0;
}

static int
pusharg(const char *a)
{
	size_t l = strlen(a) + 1;   // including nul

	if (buflen >= argmax - l - argsresv || argslen + 1 >= argscap)
		return 0;

	if (buflen + l > bufcap) {
		while (buflen + l > bufcap)
			bufcap *= 2;
		buf = realloc(buf, bufcap);
		if (!args)
			exit(1);
	}

	memcpy(buf + buflen, a, l);
	buflen += l;
	argslen++;

	return 1;
}

static int
run()
{
	pid_t pid, lpid;

	while (runjobs >= maxjobs)
		mywait();
	runjobs++;
	iterations++;

	scanargs();

	if (!args[0]) {
		fprintf(stderr, "xe: no command\n");
		exit(126);
	}

	if (nflag) {
		trace();
		runjobs--;
		return 0;
	}

	int pipefd[2];
	if (Lflag) {
		if (pipe(pipefd) < 0)
			exit(126);
	}

	unsigned char status;
	int alivepipefd[2];
	if (pipe(alivepipefd) < 0)
		exit(126);
	fcntl(alivepipefd[0], F_SETFD, FD_CLOEXEC);
	fcntl(alivepipefd[1], F_SETFD, FD_CLOEXEC);

	pid = fork();
	if (pid == 0) {  // in child
		close(alivepipefd[0]);

		char iter[32];
		snprintf(iter, sizeof iter, "%ld", iterations);
		setenv("ITER", iter, 1);
		// redirect stdin to /dev/null when we read arguments from it
		// just close it if this fails
		if (input == stdin) {
			int fd = open("/dev/null", O_RDONLY);
			if (fd >= 0) {
				if (dup2(fd, 0) != 0)
					close(0);
				close(fd);
			} else {
				close(0);
			}
		}

		if (Lflag) {
			if (dup2(pipefd[1], 1) < 0)
				exit(126);
			close(pipefd[0]);
			close(pipefd[1]);
		}

		if (qflag) {
			int fd = open("/dev/null", O_RDONLY);
			if (fd >= 0) {
				dup2(fd, 2);
				dup2(fd, 1);
				close(fd);
			}
		}

		execvp(args[0], args);

		status = (errno == ENOENT ? 127 : 126);
		if (write(alivepipefd[1], &status, 1) != 1) {
			/* ignored */
		}
		fprintf(stderr, "xe: %s: %s\n", args[0], strerror(errno));
		_exit(status);
	} else if (pid < 0) {  // fork failed
		fprintf(stderr, "xe: %s: %s\n", args[0], strerror(errno));
		exit(126);
	}

	close(alivepipefd[1]);
	if (read(alivepipefd[0], &status, 1) == 1) {
		if (status == 126)
			exit(126);
		if (status == 127)
			exit(127);
	}
	close(alivepipefd[0]);

	if (Lflag) {
		long iter = iterations;
		lpid = fork();
		if (lpid == 0) {  // in line-logging child
			char *line = 0;
			size_t linelen = 0;

			close(0);
			close(pipefd[1]);
			FILE *input = fdopen(pipefd[0], "r");
			if (!input)
				exit(126);

			setvbuf(stdout, 0, _IOLBF, 0);

			while (1) {
				int rd = getdelim(&line, &linelen, '\n', input);
				if (rd == -1)
					exit(0);

				if (vflag > 1 || Lflag > 1)
					printf("%04ld= ", iter);
				fwrite(line, 1, rd, stdout);
			}
		}
		if (lpid < 0)
			exit(126);

		close(pipefd[0]);
		close(pipefd[1]);
	}

	if (vflag) {
		if (vflag > 1)
			fprintf(traceout, "%04ld> ", iterations);
		trace();
	}

	int i;
	for (i = 0; i < maxjobs; i++)
		if (!children[i].pid) {
			children[i].pid = pid;
			children[i].iter = iterations;
			break;
		}

	return 0;
}

void
toolong()
{
	fprintf(stderr, "xe: fixed argument list too long\n");
	exit(1);
}

int
parse_jobs(char *s)
{
	char *e;
	int n;

	if (*s && s[strlen(s) - 1] == 'x') {
#ifdef _SC_NPROCESSORS_ONLN
		n = (int)sysconf(_SC_NPROCESSORS_ONLN);
#else
		n = 1;
#endif
		double d = 0.0;
		errno = 0;
		d = strtod(s, &e);
		if (errno != 0 || *e != 'x' || d <= 0) {
			fprintf(stderr, "xe: can't parse multiplier '%s'.\n", s);
			exit(1);
		}
		n = (int)(d * n);
		if (n < 1)
			n = 1;
	} else if (strcmp(s, "j") == 0) {
		n = -1;
	} else {
		errno = 0;
		n = strtol(s, &e, 10);
		if (errno != 0 || *e) {
			fprintf(stderr, "xe: can't parse number '%s'.\n", s);
			exit(1);
		}
	}

#ifdef _SC_NPROCESSORS_ONLN
	if (n <= 0)
		n = (int)sysconf(_SC_NPROCESSORS_ONLN);
#endif
	if (n <= 0)
		n = 1;

	return n;
}

// perc is a variant of fnmatch, implemented here using left derivatives:
//
// differences to fnmatch:
// ** matches any substring
// % matches any nonempty substring (longest first), is kept for substitution
// [!...] and [^...] are supported
// [] matches nothing
// [...] never matches /
// {a,b,c} alternation
// {} matches empty string
// no special prefix . handling: *le matches .profile
// no collations in []

size_t perc_len;
char *perc_str;

char *
perc(char *pat, char *str, int lvl)
{
	char *s, *e;
	int neg, matched;
	size_t l;

	switch (*pat) {
	case '\0':
		if (lvl > 0) {
			fprintf(stderr, "xe: unmatched { in glob\n");
			exit(1);
		}
		return *str ? 0 : str;
	case '?':
		if (!*str || *str == '/')
			return 0;
		return perc(pat+1, str+1, lvl);
	case '/':
		if (*str != '/')
			return 0;
		while (*pat == '/')
			pat++;
		while (*str == '/')
			str++;
		return perc(pat, str, lvl);
	case '*':
		pat++;
		if (*pat == '*') {  // any substring
			while (*pat == '*')
				pat++;
			l = strlen(str) + 1;
		} else {  // any substring but no more than the next /
			if ((e = strchr(str, '/')))
				l = e - str + 1;
			else
				l = strlen(str) + 1;
		}
		while (l--)
			if ((s = perc(pat, str+l, lvl)))
				return s;
		return 0;
	case '%':
		// any nonempty substring
		for (l = strlen(str); l >= 1; l--)
			if ((s = perc(pat+1, str+l, lvl))) {
				perc_str = str;
				perc_len = l;
				return s;
			}
		return 0;
	case '[':
		if (!*str || *str == '/')
			return 0;
		pat++;
		neg = 0;
		if (*pat == '^' || *pat == '!') {
			pat++;
			neg = 1;
		}
		for (matched = 0; *pat && *pat != ']'; pat++) {
			if (*pat == '\\')
				pat++;
			if (pat[1] == '-' && pat[2] != ']') {
				if (pat[0] <= *str && *str <= pat[2])
					matched = 1;
				pat += 2;
			} else if (*str == *pat) {
				matched = 1;
			}
		}
		if (*pat != ']') {
			fprintf(stderr, "xe: unmatched [ in glob\n");
			exit(1);
		}
		return (neg ^ matched) ? perc(pat+1, str+1, lvl) : 0;
	case '{':
		e = 0;
		while (*pat++ != '}') {
			if (!e)
				e = perc(pat, str, lvl+1);
			// skip to next , or }
			for (l = 0;
			     *pat && !(l == 0 && (*pat == ',' || *pat == '}'));
			     pat++)
				if (*pat == '{')
					l++;
				else if (*pat == '}')
					l--;
				else if (*pat == '[')
					while (*pat && *pat != ']') {
						if (*pat == '\\' && pat[1])
							pat++;
						pat++;
					}
		}
		return e ? perc(pat, e, lvl) : 0;
	case ',':
	case '}':
		if (lvl > 0)
			return str;
		/* FALL THROUGH */
	default:
		return (*pat == *str) ? perc(pat+1, str+1, lvl) : 0;
	}
}

int
perc_match(char *pat, char *arg)
{
	if (!strchr(pat, '/') && !strstr(pat, "**")) {
		char *d = strrchr(arg, '/');
		if (d)
			arg = d + 1;
	}

	perc_len = 0;
	perc_str = 0;

	return perc(pat, arg, 0) != 0;
}

char *
perc_subst(char *pat, char *base, char *arg)
{
	static char buf[2048];
	size_t l;
	char *dir = base;

	if (strcmp(arg, replace) == 0)
		return base;

	if (!strchr(pat, '/') && !strstr(pat, "**")) {
		char *d = strrchr(base, '/');
		if (d)
			base = d + 1;
	}

	char *t = strchr(arg, '%');

	if (!t)
		return arg;

	if (perc_len)
		l = snprintf(buf, sizeof buf, "%.*s%.*s%.*s%.*s",
		    (int)(base - dir),
		    dir,

		    (int)(t - arg),
		    arg,

		    (int)perc_len,
		    perc_str,

		    (int)(arg + strlen(arg) - t),
		    t+1);
	else
		l = snprintf(buf, sizeof buf, "%.*s%.*s%s%.*s",
		    (int)(base - dir),
		    dir,

		    (int)(t - arg),
		    arg,

		    base,

		    (int)(arg + strlen(arg) - t),
		    t+1);

	if (l >= sizeof buf) {
		fprintf(stderr, "xe: result of percent substitution too long\n");
		exit(1);
	}

	return buf;
}

static void
trace_to_stderr()
{
	FILE *linestderr = fdopen(2, "w");
	if (linestderr) {
		setvbuf(linestderr, 0, _IOLBF, 0);
		traceout = linestderr;
	} else {
		traceout = stderr;
	}
}

int
main(int argc, char *argv[], char *envp[])
{
	int c, i, j, cmdend;
	char *arg;

	bufcap = 4096;
	buf = malloc(bufcap);

	argscap = 8192;
	args = malloc(sizeof args[0] * argscap);

	if (!buf || !args)
		exit(1);

	argmax = sysconf(_SC_ARG_MAX);
	while (*envp)  // subtract size of environment
		argmax -= strlen(*envp++) + 1 + sizeof *envp;
	argmax -= 4 * 1024;  // subtract 4k for safety
	if (argmax > 128 * 1024)  // upper bound
		argmax = 128 * 1024;
	if (argmax <= 0)  // lower bound
		argmax = _POSIX_ARG_MAX;

	traceout = stdout;

	while ((c = getopt(argc, argv, "+0A:FI:LN:Raf:j:npqs:v")) != -1)
		switch (c) {
		case '0': delim = '\0'; break;
		case 'A': argsep = optarg; Aflag++; break;
		case 'I': replace = optarg; break;
		case 'L': Lflag++; break;
		case 'N': maxatonce = atoi(optarg); break;
		case 'F': Fflag++; break;
		case 'R': Rflag++; break;
		case 'a': aflag++; break;
		case 'f': fflag = optarg; break;
		case 'j': maxjobs = parse_jobs(optarg); break;
		case 'n': nflag++; break;
		case 'p': pflag++; break;
		case 'q': qflag++; break;
		case 's': sflag = optarg; break;
		case 'v': vflag++; break;
		default:
			fprintf(stderr,
			    "Usage: %s [-0FLRnqv] [-I arg] [-N maxargs] [-j maxjobs] COMMAND...\n"
			    "     | -p PATTERN COMMAND... [+ PATTERN COMMAND...]...\n"
			    "     | -f ARGFILE COMMAND...\n"
			    "     | -s SHELLSCRIPT\n"
			    "     | -a COMMAND... -- ARGS...\n"
			    "     | -A ARGSEP COMMAND... ARGSEP ARGS...\n",
			    argv[0]);
			exit(1);
		}

	children = calloc(sizeof (struct child), maxjobs);
	if (!children)
		exit(1);

	if (!Lflag || vflag <= 1)
		trace_to_stderr();

	if (aflag || Aflag) {
		input = 0;
	} else if (!fflag || strcmp("-", fflag) == 0) {
		input = stdin;
	} else {
		input = fopen(fflag, "rb");
		if (!input) {
			fprintf(stderr, "xe: opening %s: %s\n",
			    fflag, strerror(errno));
			exit(1);
		}
		fcntl(fileno(input), F_SETFD, FD_CLOEXEC);
	}

	cmdend = argc;
	if (aflag) {  // find first -- in argv
		for (i = 1; i < argc; i++)
			if (strcmp(argv[i], "--") == 0) {
				inargs = argv + i+1;
				cmdend = i;
				break;
			}
		if (sflag) {  // e.g. on xe -s 'echo $1' -a 1 2 3
			cmdend = optind;
			inargs = argv + cmdend;
		}
	} else if (Aflag) {  // find first argsep after optind
		for (i = optind; i < argc; i++) {
			if (strcmp(argv[i], argsep) == 0) {
				inargs = argv + i+1;
				cmdend = i;
				break;
			}
		}
		if (i >= argc) {
			fprintf(stderr, "xe: '-A %s' used but no separator "
			    "'%s' found in command line.\n", argsep, argsep);
			exit(1);
		}
	}

	int keeparg = 0;

	if (pflag) {
		if (maxatonce != 1) {
			fprintf(stderr,
			    "xe: -p cannot be used together with -N.\n");
			exit(1);
		}

		while (1) {
			buflen = 0;
			argslen = 0;

			while (runjobs >= maxjobs)
				mywait();
			if (!(arg = getarg()))
				break;

			int n;
			for (n = optind, i = n + 1; n < cmdend; n = i + 1) {
				char *pat = argv[n];

				int matched = perc_match(pat, arg);

				if (sflag && matched) {
					pusharg("/bin/sh");
					pusharg("-c");
					pusharg(sflag);
					pusharg("/bin/sh");
				}

				for (i = n + 1; i < cmdend; i++) {
					if (argv[i][0] == '+' &&
					    argv[i][1] == '\0')
						break;

					if (matched &&
					    !pusharg(perc_subst(pat, arg, argv[i])))
						toolong();
				}

				if (matched) {
					run();
					break;
				}
			}
		}
	} else {
		while (1) {
			// check if there is an arg from a previous iteration
			if (!keeparg) {
				while (runjobs >= maxjobs)
					mywait();

				arg = getarg();
				if (!arg)
					break;
			}
			keeparg = 0;

			buflen = 0;
			argslen = 0;

			if (sflag) {
				pusharg("/bin/sh");
				pusharg("-c");
				pusharg(sflag);
				pusharg("/bin/sh");
			} else if (optind >= cmdend) {
				pusharg("printf");
				pusharg("%s\\n");
			}

			for (i = optind; i < cmdend; i++) {
				if (*replace && strcmp(argv[i], replace) == 0)
					break;
				if (!pusharg(argv[i]))
					toolong();
			}

			if (!pusharg(arg))
				toolong();
			i++;

			// reserve space for final arguments
			for (argsresv = 0, j = i; j < cmdend; j++)
				argsresv += 1 + strlen(argv[j]);

			// fill with up to maxatonce arguments
			for (j = 0; maxatonce < 1 || j < maxatonce-1; j++) {
				arg = getarg();
				if (!arg)
					break;
				if (!pusharg(arg)) {
					// we are out of space,
					// deal with this arg in the next iter
					keeparg = 1;
					break;
				}
			}

			for (argsresv = 0, j = i; j < cmdend; j++)
				if (!pusharg(argv[j]))
					toolong();

			run();
		}
	}

	while (mywait())
		;

	free(buf);
	free(args);
	free(line);

	if (Rflag && iterations == 0)
		return 122;
	if (failed)
		return 123;
	return 0;
}
